internal/states/statecrypto/implementations/aes256state/aes256state.go

```diff
@@ -0,0 +1,150 @@
+	validator := regexp.MustCompile("^[0-9a-f]{64}$")
+	if !validator.MatchString(hexKey) {
+		return []byte{}, fmt.Errorf("key was not a hex string representing 32 bytes, must match [0-9a-f]{64}")
+	}
+
+	key, _ := hex.DecodeString(hexKey)
+
+	return key, nil
+}
+
+func (a *AES256StateWrapper) parseKeyFromConfiguration(config cryptoconfig.StateCryptoConfig) error {
+	hexkey, ok := config.Parameters["key"]
+	if !ok {
+		return fmt.Errorf("configuration for AES256 needs the parameter 'key' set to a 32 byte lower case hexadecimal value")
+	}
+
+	key, err := parseKey(hexkey)
+	if err != nil {
+		return err
+	}
+
+	a.key = []byte(key)
+	return nil
+}
+
+// determine if data (which is a []byte containing a json structure) is encrypted, that is, of the following form:
+//     {"crypted":"<hex containing iv and payload>"}
+func (a *AES256StateWrapper) isEncrypted(data []byte) bool {
+	validator := regexp.MustCompile(`^{"crypted":".*$`)
+	return validator.Match(data)
+}
+
+func (a *AES256StateWrapper) isSyntacticallyValidEncrypted(data []byte) bool {
+	validator := regexp.MustCompile(`^{"crypted":"[0-9a-f]+"}$`)
+	return validator.Match(data)
+}
+
+func (a *AES256StateWrapper) attemptDecryption(jsonCryptedData []byte, key []byte) ([]byte, error) {
+	if !a.isSyntacticallyValidEncrypted(jsonCryptedData) {
+		return []byte{}, fmt.Errorf("ciphertext contains invalid characters, possibly cut off or garbled")
+	}
+
+	// extract the hex part only, cutting off {"crypted":" (12 characters) and "} (2 characters)
+	src := jsonCryptedData[12 : len(jsonCryptedData)-2]
+
+	ciphertext := make([]byte, hex.DecodedLen(len(src)))
+	n, err := hex.Decode(ciphertext, src)
+	if err != nil {
+		return []byte{}, err
+	}
+	if n != hex.DecodedLen(len(src)) {
+		return []byte{}, fmt.Errorf("did not fully decode, only read %d characters before encountering an error", n)
+	}
+
+	block, err := aes.NewCipher(key)
+	if err != nil {
+		return []byte{}, err
+	}
+
+	if len(ciphertext) < aes.BlockSize {
+		return []byte{}, fmt.Errorf("ciphertext too short, did not contain initial vector")
+	}
+	iv := ciphertext[:aes.BlockSize]
+	payloadWithHash := ciphertext[aes.BlockSize:]
+
+	stream := cipher.NewCFBDecrypter(block, iv)
+
+	// XORKeyStream can work in-place if the two arguments are the same.
+	stream.XORKeyStream(payloadWithHash, payloadWithHash)
+
+	plaintextPayload := payloadWithHash[:len(payloadWithHash)-sha256.Size]
+	hashRead := payloadWithHash[len(payloadWithHash)-sha256.Size:]
+
+	hashComputed := sha256.Sum256(plaintextPayload)
+	for i, v := range hashComputed {
+		if v != hashRead[i] {
+			return []byte{}, fmt.Errorf("hash of decrypted payload did not match at position %d", i)
+		}
+	}
+
+	// payloadWithHash is now decrypted
+	return plaintextPayload, nil
+}
+
+// Encrypt data (which is a []byte containing a json structure) into a json structure
+//      {"crypted":"<hex-encoded random iv + hex-encoded CFB encrypted data including hash>"}
```

Comment:
I am feeling slightly uneasy retrieving the decryption algorithm from the (potentially manipulated) remote state. It's probably safe, but I'd rather not. If key / algorithm rotation is desired or necessary, this can be done using the configuration pulled in from the environment variables.